src/Units/UnitModel.java

```diff
@@ -6,88 +6,115 @@
 import java.util.Set;
 import java.util.Vector;
 
-import Engine.GameAction;
 import Engine.UnitActionType;
+import Engine.GameEvents.GameEventQueue;
 import Terrain.Location;
+import Terrain.MapMaster;
 import Terrain.TerrainType;
 import Units.MoveTypes.MoveType;
-import Units.Weapons.WeaponModel;
 
 /**
  * Defines the invariant characteristics of a unit. One UnitModel can be shared across many instances of that Unit type.
  */
-public class UnitModel implements Serializable
+public abstract class UnitModel implements Serializable
 {
   private static final long serialVersionUID = 1L;
 
-  public enum UnitEnum
+  /**
+   * Defines the core types of units we expect to be in any given unit set.
+   * Unit sets should have no less than one of each.
+   * Unit sets should order models such that the most "sensible" type (generally cheapest, land) is the first type of a given role.
+   */
+  public enum UnitRoleEnum
```

Comment:
`UnitRoleEnum`s seem to have some overlap. `MECH`s are pretty clearly also `INFANTRY`, Stealths are `AIR_ASSAULT` but could also be `AIR_SUPERIORITY`, etc. Would it make sense to turn this into a bitmap? Could include things like ability to cloak/capture in there too.

Reply:
That could make sense, yeah.
I'd have to consider the implications on our existing use cases (and where we wanna go with things) a bit more before going for that, though.

Reply:
Okay. I took a look at it.
I'm actually quite liking the whole flags idea.

It does occur to me that it's definitely overlapping with our whole "chassis" concept, and that maybe the chassis should *also* be folded into this new role bitmap?
Then again, that means we need to decide what sort of semantics we want to define for how e.g. transports deal with multiple-chassis units. If a Bike is both TROOP and TANK, or an amphibious tank is both TANK and SUB, should APCs and Landers still load them (respectively)?

There's one incompatibility with current usage, and I'm not sure it's all that important based on the other comment thread about the Battleships:
I wanted it to be possible to make completely unit-set-agnostic maps by specifying unit roles rather than unit names (particularly for FTA counter inf purposes)... but I think it's fair to say that high-quality maps *in general* are balanced around a particular unit set.

This might matter a little If we do end up renaming infantry to riflemen, but then again I can't see that being too world-shattering.

Reply:
Counterpoint to the chassis thing:
The overlap I was seeing was primarily that a CO (or even AI) could be interested in using chassis or role for filtering unit types. This effect is particularly strong in `INFANTRY`/`TROOP`.
That particular case is quite strongly overloaded, but the other chassis types don't really align quite so easily.

It's also relevant that the chassis enum mixes both morphology and location in it: it would seem to be pretty incoherent for something to be `TROOP`, `SUB`, and `AIR_HIGH` all at once (and no other chassis types). (I guess you could make an argument about a Protoss Colossus standing in water, or maybe an Avatar of Khaine standing in water... but still.)

Reply:
> If a Bike is both TROOP and TANK, or an amphibious tank is both TANK and SUB, should APCs and Landers still load them (respectively)

Yes. If APCs load troops, then they should be able to load anything that has that flag.

> The overlap I was seeing

I mean, infantry and mechs are both troops. If we introduce engineers and armored suits, they would presumably also be troops. I see no problem with AIs/COs filtering by role or by chassis.

> it would seem to be pretty incoherent for something to be TROOP, SUB, and AIR_HIGH

Well, yeah. Are you just worried about people setting multiple bits that seem to contradict each other? Lets just not do that. Alternatively, if we ever decide to build an Iron Man suit, now we can.